Induction naming convention (#522)
leanprover-community/leanprover-community.github.io: templates/contribute/naming.md

and there is no intention to change this.
When such an automatically generated lemma already exists,
and a bidirectional lemma is needed, it may be named `.inj_iff`.

### Inductive and recursive principles

Inductive/recursive principles are ways to construct data or proofs for all elements of some type `T`,
by providing ways to construct this data or proof in more constrained specific contexts. 

> These principles generally are phrased to first accept a *motive*,

Reviewer: ```suggestion
These principles generally are phrased to first accept a `motive` argument,
```
since the argument is not only a motive, but should be _called_ `motive`.

Author: that isn't consistently done over all of mathlib though. for example, check out [Quotient.inductionOn₂](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Quot.html#Quotient.inductionOn%E2%82%82%27) for an example in Mathlib, or [WellFounded.induction](https://leanprover-community.github.io/mathlib4_docs/Init/WF.html#WellFounded.induction) in Init

Author: additionally, i think that at least for `Prop`-typed motives, naming the variable `P` for predicate is clearer, particularly since the term "motive" might be unfamiliar to new members. (i myself found it intimidating, as i hadn't come across the notion in my maths undergrad.) because of this, i'd rather emphasise that it is a technical term than that it is the name commonly used for such a variable.

Author: my opinion right now also extends to `Type`-typed motives, but at least `Prop` gives a clear alternative.

Reviewer: You're right, `P` probably is clearer for `Prop` than `motive`. Can you post that suggestion in the zulip thread to get consensus?

which declares what property we are proving or what data we are constructing for all `T`.
When the motive eliminates into `Prop`, it is an induction principle, and the name should contain
`induction`. On the other hand, when the motive eliminates into `Sort u` or `Type u`,
it is a recursive principle, and the name should contain `rec` instead.

Additionally, the name should contain `on` iff in the argument order, the value comes before the constructions.

The following table summarizes these naming conventions.

> |motive eliminates into: | `Prop`           | `Sort u` or `Type u`|

Reviewer: ```suggestion
| motive eliminates into: | `Prop`           | `Sort u` or `Type u` |
```
and more spaces below

|----------------------|------------------|-----------|
| value first          | `T.induction_on` | `T.recOn` |

> | constructions first  | `T.induction`    | `T.rec`   |

Reviewer: Can you fix the alignment of this table?

> Variation on these names are acceptable when neccessary (e.g. for disambiguation)

Reviewer: Should we prescrbe whether to use `fooRecOn` or `recFooOn` or `recOnFoo`? Do you know if mathlib is consistent?

Author: a quick search using loogle suggests that there doesn't seem to be too much consistency between Init, Mathlib, and the modules between. 
The main takeaway seems to be that both pre- and suffixes are used, but no infixes. (i did notice that `brecOn` seems to be a common name, but i don't know what the `b` signifies, do you know?).

When such an automatically generated lemma already exists,
and a bidirectional lemma is needed, it may be named `.inj_iff`.

> ### Inductive and recursive principles
> 
> Inductive/recursive principles are ways to construct data or proofs for all elements of some type `T`,

Reviewer: ```suggestion
### Induction and recursion principles

Induction/recursion principles are ways to construct data or proofs for all elements of some type `T`,
```

by providing ways to construct this data or proof in more constrained specific contexts. 
These principles should be phrased to accept a `motive` argument,
which declares what property we are proving or what data we are constructing for all `T`.
When the motive eliminates into `Prop`, it is an induction principle, and the name should contain
`induction`. On the other hand, when the motive eliminates into `Sort u` or `Type u`,
it is a recursive principle, and the name should contain `rec` instead.

Additionally, the name should contain `on` iff in the argument order, the value comes before the constructions.

> The following table summarizes these naming conventions.

Reviewer: ```suggestion
The following table summarizes these naming conventions:
```

| motive eliminates into: | `Prop`           | `Sort u` or `Type u` |
|-------------------------|------------------|----------------------|
| value first             | `T.induction_on` | `T.recOn`            |
| constructions first     | `T.induction`    | `T.rec`              |

> Variation on these names are acceptable when necessary (e.g. for disambiguation)

Reviewer: ```suggestion
Variation on these names are acceptable when necessary (e.g. for disambiguation).
```
